File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
from itertools import combinations
input = sys.stdin.readline

N = int(input())
S = [list(map(int, input().split())) for _ in range(N)]
#print(S)

num = [i for i in range(N)]
team = list(combinations(num, N//2))
#print(team)

min_diff = 20**9
for k in range(len(team)//2):
    teamA = team[k]
    start = 0
    for i in range(len(teamA)):
        for j in range(len(teamA)):
            start += S[teamA[i]][teamA[j]]
    #print(teamA, start)

    teamB = team[-1-k]
    link = 0
    for i in range(len(teamB)):
        for j in range(len(teamB)):
            link += S[teamB[i]][teamB[j]]
    #print(teamB, link)
    min_diff = min(min_diff, abs(start-link))

print(min_diff)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;


public class Main {
    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static int n, minDiff = Integer.MAX_VALUE;
    private static int[][] A;
    private static int[] visited;

    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());
        A = new int[n][n];
        visited = new int[n];

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < n; j++) {
                A[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        combination(0, 0);
        System.out.println(minDiff);
    }

    private static void combination(int depth, int s) {
        if (depth == n/2) {
            int start = 0;
            int link = 0;
            for (int i = 0; i < n-1; i++) {
                for (int j = i+1; j < n; j++) {
                    if (visited[i] == 1 && visited[j] == 1) {
                        start += (A[i][j] + A[j][i]);
                    }
                    else if (visited[i] == 0 && visited[j] == 0) {
                        link += (A[i][j] + A[j][i]);
                    }
                }
            }
            minDiff = Math.min(minDiff, Math.abs(start-link));
            return;
        }

        for (int i = s; i < n; i++) {
            if (visited[i] == 0) {
                visited[i] = 1;
                combination(depth+1, i+1);
                visited[i] = 0;
            }
        }

    }
}